relayer/spy_relayer/src/listener/validation.ts

```diff
@@ -105,35 +111,57 @@ export async function parseAndValidateVaa(
   //   return "VAA is not from a monitored contract.";
   // }
 
-  const isCorrectPayloadType = parsedVaa.payload[0] === 1;
-
-  if (!isCorrectPayloadType) {
-    logger.debug("Specified vaa is not payload type 1.");
-    return "Specified vaa is not payload type 1..";
+  const payloadType = parsedVaa.payload[0];
+  let parsedVaaPayload: any = null;
+  if (payloadType !== 3) {
+    const error = "Specified vaa is not swim payload"
+    logger.debug(error);
+    return error
   }
 
-  let parsedPayload: any = null;
   try {
-    parsedPayload = parseTransferPayload(Buffer.from(parsedVaa.payload));
+    parsedVaaPayload = parseTransferWithArbPayload(Buffer.from(parsedVaa.payload));
   } catch (e) {
     logger.error("Encountered error while parsing vaa payload" + e);
   }
 
-  if (!parsedPayload) {
-    logger.debug("Failed to parse the transfer payload.");
-    return "Could not parse the transfer payload.";
+  if (!parsedVaaPayload) {
+    logger.debug("Failed to parse the transfer with data payload.");
+    return "Could not parse the transfer with data payload.";
+  }
+
+  // Make sure this is specifically a Swim Payload 3
+  let swimPayload: any = null;
+  try {
+    swimPayload = parseSwimPayload(Buffer.from(parsedVaaPayload.extraPayload));
+  } catch (e) {
+    logger.error("Encountered error while parsing swim payload from arbitrary data in payload 3" + e) ;
+    return "Could not parse swim payload";
+  }
+
+  // TODO swim payload version check
+  if (swimPayload.swimMessageVersion !== 1) {
+    //logger.debug("swimMessageVersion is not 1");
+  }
+
+  // We should only be getting messages from one specific swim EVM routing contract.
+  const expectedSwimEvmContractAddress = tryNativeToHexString(env.swimEvmContractAddress, CHAIN_ID_ETH);
```

Comment:
this should be an env/config filter similar to approved tokens.

Reply:
I think you meant `SPY_SERVICE_FILTERS` instead of `SUPPORTED_TOKENS`, since the token validation also happens inside here

I looked into using `SPY_SERVICE_FILTERS` to filter on only swim routing contract VAAs. The `emitterFilter` that are passed into `subscribeSignedVAA` aren't very robust. They only understand filtering on the `emitterAddress` field on the top level VAA. For testing purposes, that means that I added the wormhole token bridge addresses for goerli and bsctestnet to the `SPY_SERVICE_FILTERS`.

Since the swim evm routing contract is in the payload senderAddress field and not the top level VAA, I can't filter on that directly with `subscribeSignedVAA`, so I put the swim evm contract check here instead

```
{
  version: 1,
  guardianSetIndex: 0,
  signatures: [...],
  ...
  emitterAddress: '0x000000000000000000000000f890982f9310df57d00f659cf4fd87e65aded8d7', // can filter on this
  payload: {
    module: 'TokenBridge',
    type: 'TransferWithPayload',
    ...
    fromAddress: '0x000000000000000000000000a33e4d9624608c468fe5466dd6cc39ce1da4ff78', // can't filter on this
    ...
  }
}
```

Reply:
you don't have to do it directly in the subscribeSignedVAA you can do a subsequent level of filtering when you're streaming the signedVAAs that did pass the SPY_SERVICE_FILTERS.
regardless you'll still want to have some type of configuration similar to SPY_SERVICE_FILTERS since you'll need to have different from_address filters depending on which chain the VAA originates from.